### PR TITLE
feat: log raw importance ratios and fraction of truncation/masking in vLLM importance sampling correction

### LINKED ISSUE #5231: Logging how vllm importance ratios are truncated/masked in GRPOTrainer
### Feature request

I think logging more values related to vllm importance ratios will be helpful, e.g. logging max/min/mean of raw importance ratios (before truncation/mask) and the fraction of importance samples being truncated/masked.

### Motivation

As of now, all the logged values related to vllm importance sampling correction are based on already truncated/masked values, so it is hard to see what the raw values look like and how many importance ratios are truncated/masked.

Where the importance ratios are truncated/masked:
https://github.com/huggingface/trl/blob/1850da5a3cbaeb043b483ae78986c43246501d40/trl/trainer/grpo_trainer.py#L1807-L1814

Where the values are logged:
https://github.com/huggingface/trl/blob/1850da5a3cbaeb043b483ae78986c43246501d40/trl/trainer/grpo_trainer.py#L1955-L1977
https://github.com/huggingface/trl/blob/1850da5a3cbaeb043b483ae78986c43246501d40/trl/trainer/grpo_trainer.py#L1990

### Your contribution

I can send a PR to add metrics using importance ratios before truncation/masking.

### PR DESCRIPTION
**Pull request by muupan**
# What does this PR do?

Resolves #5231 

This PR adds new logged metrics:
- `sampling/raw_importance_sampling_ratio/{min,max,mean}`
  - These are the same with existing `sampling/importance_sampling_ratio/{min,max,mean}` except being computed with values before truncation or masking.
- `sampling/frac_modified_importance_sampling_ratio`
  - This is the fraction of importance sampling ratio values that are either truncated or masked.

I ran the following code to verify the output:

```python
import argparse

from datasets import load_dataset
from trl import GRPOConfig, GRPOTrainer
from trl.rewards import accuracy_reward


parser = argparse.ArgumentParser()
parser.add_argument("--vllm_importance_sampling_mode", type=str, default="sequence_mask")
args = parser.parse_args()

dataset = load_dataset("trl-lib/DeepMath-103K", split="train[:5]")

args = GRPOConfig(
    output_dir=f"outputs/{args.vllm_importance_sampling_mode}",
    vllm_importance_sampling_correction=True,
    vllm_importance_sampling_mode=args.vllm_importance_sampling_mode,
    vllm_importance_sampling_cap=1.2,
    use_vllm=True,
    vllm_mode="colocate",
    max_steps=5,
    num_train_epochs=1,
    logging_steps=1,
    save_strategy="no",
    report_to="tensorboard",
)

trainer = GRPOTrainer(
    args=args,
    model="Qwen/Qwen2-0.5B-Instruct",
    reward_funcs=accuracy_reward,
    train_dataset=dataset,
)
trainer.train()
```

```
uv run --no-sync accelerate launch --num_processes 1 train_grpo_example.py --vllm_importance_sampling_mode token_truncate
uv run --no-sync accelerate launch --num_processes 1 train_grpo_example.py --vllm_importance_sampling_mode token_mask
uv run --no-sync accelerate launch --num_processes 1 train_grpo_example.py --vllm_importance_sampling_mode sequence_truncate
uv run --no-sync accelerate launch --num_processes 1 train_grpo_example.py --vllm_importance_sampling_mode sequence_mask
```

From the tensorboard records, you can see:
- The new metrics are all recorded.
- `sampling/raw_importance_sampling_ratio/max` are higher than the cap value of 1.2, while `sampling/importance_sampling_ratio/max` is upper bounded by it, which is expected as the latter are affected by truncation or masking. Masking leads to lower values than truncation, which is also expected.
- `sampling/frac_modified_importance_sampling_ratio` is higher with sequence-level IS than with token-level IS.

Screenshot 2026-03-09 at 0 49 18

Screenshot 2026-03-09 at 0 49 05

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

---

> [!NOTE]
> **Low Risk**
> Low risk: changes are limited to additional metric logging around vLLM importance-sampling correction, without altering the loss or training behavior beyond minor extra tensor ops.
> 
> **Overview**
> Adds additional GRPO vLLM importance-sampling diagnostics by tracking the *raw* (pre-cap) importance sampling ratios alongside the existing capped/masked ratios.
> 
> `GRPOTrainer` now logs `sampling/raw_importance_sampling_ratio/{min,mean,max}` and `sampling/frac_modified_importance_sampling_ratio` (fraction of ratios changed by truncation/masking), and the GRPO docs are updated to list these new metrics.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 32a1746dcd83c9287e0eba2bac881dd7bc60fcb0. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by qgallouedec** [comment]
thanks! can you add a mention of this in grpo_trainer.md as well

**Comment by muupan** [comment]
It seems that grpo_trainer.md does not mention `sampling/*` metrics at all. I added mentions to them including existing ones.

**Comment by muupan** [comment]
@qgallouedec Gentle ping on this PR when you have a chance.

**Comment by qgallouedec** [comment]
Thanks for this @muupan, and sorry for the long delay getting to it!

The raw-ratio and frac-modified diagnostics are genuinely useful and still aren't on main. Since you opened this, though, the vLLM IS clipping was reworked from a single `vllm_importance_sampling_cap` into a two-sided `vllm_importance_sampling_clip_min` /` vllm_importance_sampling_clip_max` (the old cap is now deprecated), so the PR conflicts and the docs you added still refer to `vllm_importance_sampling_cap`.

Would you be up for rebasing onto main and adapting to the two-sided clip? The new `invalid_mis_mask` in the mask branch should make the `frac_modified` computation straightforward, and the doc bullets just need the cap → clip_min/clip_max rename. If you're no longer able to, no worries, let us know and we'll pick it up or close it out.

**Comment by muupan** [comment]
Ok, I'll update it accordingly.
